feat(submit_transaction) : refactor submit transaction command

cain4chain opened the pull request:
## Description of Changes

chore(submit_transaction) : add error msg and x-topics header
chore(submit_transaction_handler): add data and type validaiton

## Linked Issues / Tickets

Closes : [REFACTOR] Refine the NewSubmitTransactionCommandHandler implementation #55

## Testing Procedure

chore(submit_transaction_handler_test) : add cases for success, http type, invalid, missing and nil

- [x] I have added new unit tests
- [x] All tests pass locally
- [x] I have tested manually in my local environment

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have run the linter

cain4chain committed:
feat(submit_transaction) : refactor submit transaction command

chore(submit_transaction) : add error msg and x-topics header
chore(submit_transaction_handler): add data and type validaiton
choresubmit_transaction_handler_test) : add cases for success, http type, invalid, missing and nil

github-actions[bot] committed:
chore: update generated files

mgosek-4chain commented on pkg/server/app/commands/submit_transaction.go:
```diff
@@ -2,13 +2,27 @@ package commands
+	// Header keys
+	headerTopics = "x-topics"
+)
+
```
```
const XTopicsHeader = "x-topics"

const RequestBodyLimit1GB = 1000 * 1024 * 1024

var (
	ErrMissingXTopicsHeader       = errors.New("missing x-topics header")
	ErrInvalidXTopicsHeaderFormat = errors.New("invalid x-topics header format")
	ErrInvalidHTTPMethod          = errors.New("invalid HTTP method")
	ErrRequestBodyRead            = errors.New("failed to read request body")
	ErrRequestBodyTooLarge        = errors.New("request body too large")
)
```

mgosek-4chain commented on pkg/server/app/commands/submit_transaction.go:
```diff
@@ -35,14 +47,76 @@ type SubmitTransactionHandler struct {
+		responseSent = true
+		jsonutil.SendHTTPResponse(w, http.StatusOK, SubmitTransactionHandlerResponse{Steak: steak})
+	}
 }
```
First and foremost, the current code has a resource leak (also present in other handlers - my fault 😅) due to not closing the response body after processing. Accumulated unclosed response bodies can degrade performance, cause memory leaks by holding unused file descriptors, and introduce other defects. 

To fix this, we need to add `defer res.Body.Close()` as the first line in the handler. 

Another issue is reading the entire request payload into a `byte` slice without setting a request body limit. This can be easily fixed using `io.LimitReader`. However, the downside of this approach is that it loads the entire request body into memory. If multiple users send requests to our server simultaneously, we could quickly run out of RAM 😿 . That's why we should process the request body in chunks. 

The original `overlay-express` sets this limit to `1GB` - let's use that as a reference. 

```
type SubmitTransactionHandler struct {
	provider         SubmitTransactionProvider
	requestBodyLimit int64
}
```

As an additional improvement, I strongly recommend extracting `overlay.TaggedBEEF` creation steps into a separate function to encapsulate everything in one place. 

```
func (s *SubmitTransactionHandler) CreateTaggedBEEFFromRequest(r *http.Request) (*overlay.TaggedBEEF, error) {
	header := r.Header.Get(XTopicsHeader)
	if header == "" {
		return nil, ErrMissingXTopicsHeader
	}

	var topics []string
	if err := json.Unmarshal([]byte(header), &topics); err != nil {
		return nil, ErrInvalidXTopicsHeaderFormat
	}

	reader := io.LimitReader(r.Body, s.requestBodyLimit)
	buff := make([]byte, 64*1024) // typically it's a best balance between performance and memory usage.
	bytesRead := 0
	for {
		n, err := reader.Read(buff)
		bytesRead += n

		if bytesRead > int(s.requestBodyLimit) {
			return nil, ErrRequestBodyTooLarge
		}

		if err == io.EOF {
			break
		}

		if err != nil {
			return nil, ErrRequestBodyRead
		}
	}

	return &overlay.TaggedBEEF{Beef: buff, Topics: topics}, nil
}
```

The handler code looks much simpler now. The last part is ensuring that the callback function is called when the `steak` is ready. The issue lies in the design of the `overlay-engine`, where we don't know whether the callback function will be called asynchronous or synchronous. 

A weak point of `http.ResponseWriter` is that it's not thread-safe, which can lead to unpredictable behavior. Instead, we should use a channel, as it ensures only a single coroutine writes to `w`. Additionally, to avoid deadlocks or hanging requests, we can use a timeout with a custom value. 

```
func (s *SubmitTransactionHandler) Handle(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	if r.Method != http.MethodPost {
		http.Error(w, ErrInvalidHTTPMethod.Error(), http.StatusMethodNotAllowed)
		return
	}

	taggedBEEF, err := s.CreateTaggedBEEFFromRequest(r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	steakChan := make(chan *overlay.Steak, 1)
	_, err = s.provider.Submit(r.Context(), *taggedBEEF, engine.SubmitModeCurrent, func(steak *overlay.Steak) {
		steakChan <- steak
	})
	if err != nil {
		jsonutil.SendHTTPResponse(w, http.StatusInternalServerError, err.Error())
		return
	}

	select {
	case steak := <-steakChan:
		jsonutil.SendHTTPResponse(w, http.StatusOK, SubmitTransactionHandlerResponse{Steak: *steak})
	case <-time.After(5 * time.Second): // This select case can be refined later, for testing purposes, let's leave it as it is.
		http.Error(w, http.StatusText(http.StatusRequestTimeout), http.StatusRequestTimeout)
	}
}
```

```
func NewSubmitTransactionCommandHandler(provider SubmitTransactionProvider) *SubmitTransactionHandler {
	if provider == nil {
		panic("submit transaction provider is nil")
	}
	return &SubmitTransactionHandler{
		provider:         provider,
		requestBodyLimit: RequestBodyLimit1GB,
	}
}
```


After all, please apply the suggestions described above. Additionally, adjust the unit tests and add any missing ones. For example, verify if the API behaves correctly when attempting to exceed the request payload.

cain4chain replied:
Thanks for all the details, I will update this as requested.

cain4chain committed:
refactor(submit_transaction) : add changes as per review

chore(submit_transaction) : add body size validation logic and byte chunk validation for memory optimization.
chore(CreateTaggedBEEFFromRequest) : extract data handling into own method
chore(submit_transaction) : improve syntax and description refinement
chore(submit_transaction_test) : Update test cases, create middleware version method to validate body size validations, add test for timeout validation using non callback handler

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
Merge branch 'main' into REFACTOR-Refine-the-NewSubmitTransactionCommandHandler

cain4chain committed:
Merge branch 'REFACTOR-Refine-the-NewSubmitTransactionCommandHandler' of https://github.com/4chain-ag/go-overlay-services into REFACTOR-Refine-the-NewSubmitTransactionCommandHandler

ac4ch commented on pkg/server/app/commands/submit_transaction.go:
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
+		if err := r.Body.Close(); err != nil {
+			fmt.Printf("Failed to close request body: %v\n", err)
+		}
+	}()
```
```suggestion
	defer defer func() {
			_ = r.Body.Close()
		}()
```
this will pass linter

cain4chain replied:
ok thanks 👍

ac4ch commented on pkg/server/app/commands/submit_transaction_test.go:
```diff
@@ -1,45 +1,253 @@
+func createTestHandlerWithLimit(provider commands.SubmitTransactionProvider, limit int64) http.HandlerFunc {
+	handler, err := commands.NewSubmitTransactionCommandHandler(provider)
+	if err != nil {
+		panic(err)
```
```suggestion
		panic(err)
```
I think now after refactoring commans and queries ... we avoiding of `panic` method .. maybe change function signature and return `(http.HandlerFunc, error)`

ac4ch commented on pkg/server/app/commands/submit_transaction.go:
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
+		return nil, ErrInvalidXTopicsHeaderFormat
+	}
+
+	reader := io.LimitReader(r.Body, s.requestBodyLimit)
```
```suggestion
	reader := io.LimitReader(r.Body, s.requestBodyLimit)
	defer func(){
	    _ =  r.Body.Close()
	}
```

ac4ch commented on pkg/server/app/commands/submit_transaction.go:
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
+	buff := make([]byte, 64*1024) // typically it's a best balance between performance and memory usage (loop and size).
+	bytesRead := 0
+	for {
+		n, err := reader.Read(buff)
```
i think  the `buff` slice is overwritten by each call of `reader.Read(buff)` 
```go
n, err := reader.Read(buff)
if n > 0 {
    if _, writeErr := beefBuffer.Write(buff[:n]); writeErr != nil {
        return nil, ErrRequestBodyRead
    }
}
``` 
this would save every chunk of the body into a buffer `beefBuffer`

cain4chain replied:
yes you are right, thanks I will update this

ac4ch commented on pkg/server/app/commands/submit_transaction.go:
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
+	}
+
+	var topics []string
+	if err := json.Unmarshal([]byte(header), &topics); err != nil {
```
the header is slice ? `X-Topics: ["topic1","topic2","topic3"]` would not be better to have `X-Topics: topic1,topic2,topic3` and split this via comma ? 
```go
header := r.Header.Get(XTopicsHeader)
if header == "" {
    return nil, ErrMissingXTopicsHeader
}

topics := strings.Split(header, ",")
```

if we have control of how header is configured :)

ac4ch commented on pkg/server/app/commands/submit_transaction.go:
```diff
@@ -29,30 +54,93 @@ type SubmitTransactionProvider interface {
+	select {
+	case steak := <-steakChan:
+		jsonutil.SendHTTPResponse(w, http.StatusOK, SubmitTransactionHandlerResponse{Steak: *steak})
+	case <-time.After(5 * time.Second): // Timeout to prevent hanging requests
```
5 seconds is okay for now, but you might want to make it configurable (like s.steakResponseTimeout) ?

cain4chain replied:
Ok will create a public function for this and make it 5 seconds for default

cain4chain committed:
chore(submit_transaction) : update as per review

chore(submit_transaction) : fix byte chunk issue related to non accumulation method.
chore(submit_transaction): Optimized syntax and add customizable timeout value
chore(submit_transaction_test) : update test file with required changes of submit_transaction.go

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
chore(test_settimeout) : update test funciton

chore(test_settimeout) : update  test settimeout function using the no callback provider

mgosek-4chain committed:
refactor(#61): add code review suggestions.

mgosek-4chain reviewed (commented):
I've added suggestions how we can improve the current code base, please take a look :) 

https://github.com/4chain-ag/go-overlay-services/tree/REFACTOR-Refine-the-NewSubmitTransactionCommandHandler-cr-suggestions

mgosek-4chain committed:
refactor(#61): add a fix for handling cases where the request payload size is exceeded.

cain4chain committed:
REFACTOR : Update code from suggestions

chore(testutil) : add testutil file for mock methods used in test files
chore(lint) : fix some lint issues

github-actions[bot] committed:
chore: update generated files

cain4chain committed:
Merge branch 'main' into REFACTOR-Refine-the-NewSubmitTransactionCommandHandler

cain4chain committed:
REFACTOR : submit transaction command handler

fix(lint) : fix line auto line spacing issue CLRF

cain4chain committed:
Merge branch 'REFACTOR-Refine-the-NewSubmitTransactionCommandHandler' of https://github.com/4chain-ag/go-overlay-services into REFACTOR-Refine-the-NewSubmitTransactionCommandHandler

mgosek-4chain commented on pkg/server/app/commands/submit_transaction.go:
```diff
@@ -29,30 +55,134 @@ type SubmitTransactionProvider interface {
-	steak, err := s.provider.Submit(r.Context(), overlay.TaggedBEEF{}, engine.SubmitModeCurrent, func(steak *overlay.Steak) {})
+	defer func() {
+		_ = r.Body.Close()
+	}()
```
This can be removed because we're already using a `SafeHandler` that handles closing the request bodies :)

cain4chain replied:
Thanks, I will update this.

cain4chain committed:
REFEACTOR : refine submit transaction command

chore(handle) : remove duplicate body close

cain4chain committed:
Merge branch 'main' into REFACTOR-Refine-the-NewSubmitTransactionCommandHandler
